avro.schema.literal support and some related patches

lxynov opened the pull request:
Closes #5001 
A related issue: https://github.com/prestodb/presto/issues/9116

jiegzhan commented:
@lxynov, you have anything to add to this PR? Thanks.

lxynov commented:
@findepi Could you help review this? Some commits are kind of hacky (due to Hive's lack of spec and doc). Feel free to slack me if you'd like to discuss in DMs.

jiegzhan commented:
@lxynov @findepi Hey guys, any progress on this PR? Thanks.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java:
```diff
@@ -358,7 +358,9 @@ private void alterTable(HiveIdentity identity, String databaseName, String table
-        if (isAvroTableWithSchemaSet(table)) {
+        // If a partition's Serde doesn't use metastore for schema (like Avro), we use table-level schema. Data publishers
+        // are responsible for making sure that table-level schema is workable for all partitions.
+        if (!isPartitionSerdesUsingMetastoreForSchema(partition)) {
```
Can you please add a test (product test?) for partitioned tables where table and partition schema mismatches?

for example

- table is ORC, partition is Parquet
- table is ORC, partition is Avro with schema url
- table is ORC, partition is Avro with schema literal
- table is ORC, partition is CSV

lxynov replied:
I was trying to add such a product test (using HQLs) which fails before the commit and succeeds after the commit, but failed to do so. But I think this commit logically makes sense and it has been deployed at our company for a long time. Please let me know if you have concerns.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -904,10 +914,16 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         String avroSchemaUrl = getAvroSchemaUrl(tableMetadata.getProperties());
+        String avroSchemaLiteral = getAvroSchemaLiteral(tableMetadata.getProperties());
         if (avroSchemaUrl != null) {
+            checkState(avroSchemaLiteral == null, "avro_schema_url and avro_schema_literal can not both be set");
```
can not -> cannot ?

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1024,6 +1040,17 @@ private String validateAndNormalizeAvroSchemaUrl(String url, HdfsContext context
+            new Schema.Parser().parse(avroSchemaLiteral);
+            return avroSchemaLiteral;
+        }
+        catch (SchemaParseException e) {
```
Can there be any other exception flying here?

lxynov replied:
By looking at the implementation of `Schema.Parser::parse()`, I think `SchemaParseException` is sufficient here

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1270,8 +1298,8 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
-        if (getAvroSchemaUrl(tableMetadata.getProperties()) != null) {
-            throw new PrestoException(NOT_SUPPORTED, "CREATE TABLE AS not supported when Avro schema url is set");
+        if (getAvroSchemaUrl(tableMetadata.getProperties()) != null || getAvroSchemaLiteral(tableMetadata.getProperties()) != null) {
+            throw new PrestoException(NOT_SUPPORTED, "CREATE TABLE AS not supported when Avro schema url or literal is set");
```
I think the reason not to support CTAS with schema url is because we need the table to be in metastore in order to get the information what the schema actually (at least this is the current implementation).

For schema literal we could obtain this information locally. Even if we do not do this (which I am fine with), we should keep those cases separate here -- as separate `if`s.

lxynov replied:
Separated `if`s

findepi commented:
> avro.schema.literal support and some related patches

 the table/partition schema handliung changes require some improvement
would it be possible to separate `avro.schema.literal` into its own PR, or are there too many dependencies?

lxynov commented:
Hi @ebyhr @findepi , this PR is ready for another review. Please let me know if you have questions around any of the commits.

ebyhr commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestAvroSchemaLiteral.java:
```diff
@@ -0,0 +1,69 @@
+    @Test(groups = AVRO)
+    public void testHiveCreatedTable()
+    {
+        onHive().executeQuery("DROP TABLE IF EXISTS test_avro_schema_literal_hive");
```
It would be better to limit `onHive()` usage for Hive specific operation because it's slow. Other places are also the same.

lxynov replied:
This test is meant to test Hive created tables, so `onHive()` is used here.

ebyhr commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestAvroSchemaLiteral.java:
```diff
@@ -0,0 +1,69 @@
+public class TestAvroSchemaLiteral
+        extends HiveProductTest
+{
+    private static final String SCHEMA_LITERAL = "{\n" +
+            "  \"namespace\": \"io.prestosql.test\",\n" +
```
nit: We can add `@Language("JSON")`. `prestosql` → `trino`

ebyhr commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/MetastoreUtil.java:
```diff
@@ -147,6 +149,14 @@ private static Properties getHiveSchema(
+        // Avro reader handles schema evolution itself.
+        if (sd.getStorageFormat().getSerDe().equals(AvroSerDe.class.getName()) && tableSd.isPresent()) {
+            for (Map.Entry<String, String> param : tableSd.get().getSerdeParameters().entrySet()) {
+                schema.setProperty(param.getKey(), (param.getValue() != null) ? param.getValue() : "");
```
How about `nullToEmpty(param.getValue())`?

ebyhr reviewed (commented):
Could you add test cases about schema evolution (maybe in TestAvroSchemaEvolution)? This PR is going to close #5001, but the test looks not covered.

lxynov commented:
@ebyhr Thanks for the review! Updated.

> Could you add test cases about schema evolution (maybe in TestAvroSchemaEvolution)? This PR is going to close #5001, but the test looks not covered.

Added in `TestAvroSchemaEvolution`

lxynov commented:
@ebyhr All tests have passed. This is ready for review

lxynov commented:
@ebyhr @findepi Could you please take a look? Thanks in advance!

> would it be possible to separate avro.schema.literal into its own PR, or are there too many dependencies?

@findepi I thought about this before but found that commit `Allow creating tables with Avro schema literal` sort of depends on commit `Call getFields for serdes not using use metastore for schema`, as the later commit changed method `isAvroTableWithSchemaSet` to `isTableSerdesUsingMetastoreForSchema`

colebow commented:
:wave: @lxynov - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
